RAGchain/benchmark/dataset/dureader_robust.py

```diff
@@ -0,0 +1,117 @@
+
+        default_metrics = self.retrieval_gt_metrics + self.answer_gt_metrics \
+                          + self.answer_no_gt_ragas_metrics + self.answer_passage_metrics
+        support_metrics = default_metrics + self.retrieval_gt_ragas_metrics + self.retrieval_no_gt_ragas_metrics
+
+        if metrics is not None:
+            # Check if your metrics are available in evaluation datasets.
+            for metric in metrics:
+                if metric not in support_metrics:
+                    raise ValueError(f"You input {metric} that this dataset evaluator not support.")
+            using_metrics = list(set(metrics))
+        else:
+            using_metrics = default_metrics
+
+        super().__init__(run_all=False, metrics=using_metrics)
+
+        self.eval_size = evaluate_size
+        self.run_pipeline = run_pipeline
+
+        self.context = deepcopy(self.dataset[['id', 'context']])
+
+        if evaluate_size is not None and len(self.dataset) > evaluate_size:
+            self.qa_data = self.dataset[:evaluate_size]
+
+    def ingest(self, retrievals: List[BaseRetrieval], db: BaseDB, ingest_size: Optional[int] = None):
+        """
+        Ingest dataset to retrievals and db.
+        :param retrievals: The retrievals that you want to ingest.
+        :param db: The db that you want to ingest.
+        :param ingest_size: The number of data to ingest. If None, ingest all data.
+        If you want to use context_recall and context_precision metrics, you should ingest all data.
+        """
+
+        ingest_data = self.context
+        if ingest_size is not None:
+            # ingest size must be larger than evaluate size.
+            if ingest_size >= self.eval_size:
+                ingest_data = ingest_data[:ingest_size]
+            else:
+                raise ValueError("ingest size must be same or larger than evaluate size")
+
+        # Create passages.
+        result = ingest_data.apply(self.__make_passages, axis=1)
+        passages = result.tolist()
+
+        for retrieval in retrievals:
+            retrieval.ingest(passages)
+        db.create_or_load()
+        db.save(passages)
+
+    def evaluate(self, **kwargs) -> EvaluateResult:
+        question = self.qa_data['question'].tolist()
+        retrieval_gt = list([uuid.UUID(gt)] for gt in self.qa_data['id'])
+        answer_gt = [[answer for answer in answer_arr['text']] for answer_arr in self.qa_data['answers']]
+
+        return self._calculate_metrics(
+            questions=question,
+            pipeline=self.run_pipeline,
+            retrieval_gt=retrieval_gt,
+            answer_gt=answer_gt,
+            **kwargs
+        )
+
+    def __make_passages(self, row):
+        passage = Passage(
+            id=row['id'],
+            content=row['context'],
+            filepath=self.file_path,
+            metadata_etc={
+                'dataset url': 'https://huggingface.co/datasets/PaddlePaddle/dureader_robust/viewer/plain_text/validation'
```

Comment:
'space' at dictionary key? This is unforgiven. 
Use undber bar (_) please.... 
You can't recognize difference between two space bar, one space bar, or tab easily... Also there are empty unicode...